tiledb/vfs.py

```diff
@@ -0,0 +1,260 @@
+        :param mode str: 'rb' for opening the file to read, 'wb' to write, 'ab' to append
+        :rtype: FileHandle
+        :return: TileDB FileIO
+        :raises TypeError: cannot convert `uri` to unicode string
+        :raises ValueError: invalid mode
+        :raises: :py:exc:`tiledb.TileDBError`
+
+        """
+        return FileIO(self, uri, mode)
+
+    def close(self, file: lt.FileHandle):
+        """Closes a VFS FileHandle object
+
+        :param FileIO file: An opened VFS FileIO
+        :rtype: FileIO
+        :return: closed VFS FileHandle
+        :raises: :py:exc:`tiledb.TileDBError`
+
+        """
+        if isinstance(file, FileIO):
+            warnings.warn(
+                f"`tiledb.VFS().open` now returns a a FileIO object. Use "
+                "`FileIO.close`.",
+                DeprecationWarning,
+            )
+        file.close()
+        return file
+
+    def write(self, file: lt.FileHandle, buff: Union[str, bytes]):
+        """Writes buffer to opened VFS FileHandle
+
+        :param FileHandle file: An opened VFS FileHandle in 'w' mode
+        :param buff: a Python object that supports the byte buffer protocol
+        :raises TypeError: cannot convert buff to bytes
+        :raises: :py:exc:`tiledb.TileDBError`
+
+        """
+        if isinstance(file, FileIO):
+            warnings.warn(
+                f"`tiledb.VFS().open` now returns a a FileIO object. Use "
+                "`FileIO.write`.",
+                DeprecationWarning,
+            )
+        if isinstance(buff, str):
+            buff = buff.encode()
+        file.write(buff)
+
+    def read(self, file: lt.FileHandle, offset: int, nbytes: int):
+        """Read nbytes from an opened VFS FileHandle at a given offset
+
+        :param FileHandle file: An opened VFS FileHandle in 'r' mode
+        :param int offset: offset position in bytes to read from
+        :param int nbytes: number of bytes to read
+        :rtype: :py:func:`bytes`
+        :return: read bytes
+        :raises: :py:exc:`tiledb.TileDBError`
+
+        """
+        if isinstance(file, FileIO):
+            warnings.warn(
+                f"`tiledb.VFS().open` now returns a a FileIO object. Use "
+                "`FileIO.seek` and `FileIO.read`.",
+                DeprecationWarning,
+            )
+            return file.read(nbytes)
+
+        if nbytes == 0:
+            return b""
+
+        return file.read(offset, nbytes)
+
+    def supports(self, scheme: str) -> bool:
+        """Returns true if the given URI scheme (storage backend) is supported
+
+        :param str scheme: scheme component of a VFS resource URI (ex. 'file' / 'hdfs' / 's3')
+        :rtype: bool
+        :return: True if the linked libtiledb version supports the storage backend, False otherwise
+        :raises ValueError: VFS storage backend is not supported
+
+        """
+        if scheme == "file":
+            return True
+
+        scheme_to_fs_type = {
```

Comment:
We need a test to check whether this is exhaustive by looking at the underlying enum. Can be done in follow-up (pls create story).

Reply:
(also, hard-coding this here is not ideal -- opened SC-16386 to add an API in core we can use instead)